components-rs/bytes.rs

```diff
@@ -0,0 +1,301 @@
+    }};
+}
+
+// Insert an element in the given hashmap field.
+macro_rules! insert_hashmap {
+    ($ptr:expr, $key:expr, $value:expr, $field:ident) => {{
+        if $ptr.is_null() {
+            return;
+        }
+        let object = &mut *$ptr;
+        object.$field.insert($key, $value);
+    }};
+}
+
+// Remove an element from the given hashmap field.
+macro_rules! remove_hashmap {
+    ($ptr:expr, $key:expr, $field:ident) => {{
+        if $ptr.is_null() {
+            return;
+        }
+        let object = &mut *$ptr;
+        object.$field.remove(&$key);
+    }};
+}
+
+// Check if an element exists in the given hashmap field.
+macro_rules! exists_hashmap {
+    ($ptr:expr, $key:expr, $field:ident) => {{
+        if $ptr.is_null() {
+            return false;
+        }
+        let object = &mut *$ptr;
+        return object.$field.contains_key(&$key);
+    }};
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_set_span_service_(ptr: *mut SpanBytes, str: *mut ZendString) {
+    set_string_field!(ptr, str, service);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_set_span_name_(ptr: *mut SpanBytes, slice: *mut ZendString) {
+    set_string_field!(ptr, slice, name);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_set_span_resource_(ptr: *mut SpanBytes, slice: *mut ZendString) {
+    set_string_field!(ptr, slice, resource);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_set_span_type_(ptr: *mut SpanBytes, slice: *mut ZendString) {
+    set_string_field!(ptr, slice, r#type);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_add_span_meta_(
+    ptr: *mut SpanBytes,
+    key: *mut ZendString,
+    val: *mut ZendString,
+) {
+    insert_hashmap!(
+        ptr,
+        convert_zend_to_bytes_string(key),
+        convert_zend_to_bytes_string(val),
+        meta
+    );
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_add_CharSlice_span_meta_(
+    ptr: *mut SpanBytes,
+    key: CharSlice,
+    val: *mut ZendString,
+) {
+    insert_hashmap!(
+        ptr,
+        convert_char_slice_to_bytes_string(key),
+        convert_zend_to_bytes_string(val),
+        meta
+    );
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_del_span_meta_(ptr: *mut SpanBytes, key: *mut ZendString) {
+    remove_hashmap!(ptr, convert_zend_to_bytes_string(key), meta);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_del_CharSlice_span_meta_(ptr: *mut SpanBytes, key: CharSlice) {
+    remove_hashmap!(ptr, convert_char_slice_to_bytes_string(key), meta);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_get_span_meta_(
+    ptr: *mut SpanBytes,
+    key: CharSlice,
+) -> CharSlice<'static> {
+    if ptr.is_null() {
+        return CharSlice::empty();
+    }
+
+    let span = &mut *ptr;
+
+    let key = convert_char_slice_to_bytes_string(key);
+
+    match span.meta.get(&key) {
+        Some(value) => {
+            CharSlice::from_raw_parts(value.as_str().as_ptr().cast(), value.as_str().len())
+        }
+        None => CharSlice::empty(),
+    }
+}
```

Comment:
We have no need to duplicating these methods which do not contain ZendString here. libdatadog covers these.